stdlib/src/memory/maybe_uninitialized.mojo

```diff
@@ -0,0 +1,77 @@
+# ===----------------------------------------------------------------------=== #
+# Copyright (c) 2024, Modular Inc. All rights reserved.
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions:
+# https://llvm.org/LICENSE.txt
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ===----------------------------------------------------------------------=== #
+
+
+# TODO: Make this public when we are certain of the design.
+# TODO: Move _size into an alias when the bug https://github.com/modularml/mojo/issues/2889
+# is fixed.
+struct _MaybeUninitialized[ElementType: CollectionElement, _size: Int = 1](
+    CollectionElement
+):
+    alias type = __mlir_type[
+        `!pop.array<`, _size.value, `, `, Self.ElementType, `>`
+    ]
+    var _array: Self.type
+
+    @always_inline
+    fn __init__(inout self):
+        """The memory is now considered uninitialized."""
+        self._array = __mlir_op.`kgen.undef`[_type = Self.type]()
+
+    @always_inline
+    fn __init__(inout self, owned value: Self.ElementType):
+        """The memory is now considered initialized."""
+        self = Self()
+        self.write(value^)
+
+    @always_inline
+    fn __copyinit__(inout self, other: Self):
+        """Calling this method assumes that the memory is initialized."""
+        self = Self()
+        initialize_pointee_copy(self.unsafe_ptr(), other.assume_initialized())
+
+    @always_inline
+    fn __moveinit__(inout self, owned other: Self):
+        """Calling this method assumes that the memory is initialized."""
+        self = Self()
+        move_pointee(src=other.unsafe_ptr(), dst=self.unsafe_ptr())
+
+    @always_inline
+    fn write(inout self, owned value: Self.ElementType):
+        """Calling this method assumes that the memory is uninitialized."""
+        self.unsafe_ptr()[] = value^
```

Comment:
```suggestion
        self.unsafe_ptr().init_pointee_move(value^)
```

**Request** The `=` operator assumes that the LHS is initialized and will run the destructor on it, causing this to try to destroy uninitialized memory. `init_pointee_move()` assumes the pointee is uninitialized and so will not run the destructor.

We should add a test for this as well:

```mojo
var storage = MaybeUninitialized[ImpossibleToDestroy]()

# Should not abort
storage.write(ImpossibleToDestroy(5))
```

As a sanity check, please confirm that that test above fails with the current code 🙂

Reply:
There is already this test in my unit test file, see line 60. And this works correctly:
```mojo
def test_write_does_not_trigger_destructor():
    var a = _MaybeUninitialized[ImpossibleToDestroy]()
    a.write(ImpossibleToDestroy(42))
```
So unless I'm missing something no change is needed here. I do agree it's strange but it works.

Reply:
Something strange must be happening. I tried this simplified example which demonstrates that `ptr[] = ImpossibleToDestroy(5)` does run a destructor:

```mojo
@value
struct ImpossibleToDestroy(Movable):
    var value: Int

    fn __del__(owned self):
        abort("ABORT")


fn main():
    var ptr = UnsafePointer[ImpossibleToDestroy].alloc(1)

    ptr[] = ImpossibleToDestroy(5)
```

So if that's not in this case, that indicates `ImpossibleToDestroy` used with `UnsafeMaybeUninitialized` is not behaving as we expect. It could be a compiler bug or something else.

Do you think you might be able to work to create a minimal reproduction of the bug?

Reply:
Yep. Opened an issue here: https://github.com/modularml/mojo/issues/2951

Reply:
Thank you Gabriel! I'll resolve this thread.

Reply:
For what its worth, I'm hypothesizing that because the value is initialized to `kgen.undef`, its undefined behavior for the compiler to run the destructor, so in the spirit of "UB lets the compiler do whatever it wants", the compiler (arguably correctly) deduces it should not run a destructor on uninitialized memory and removes it.